modules/dhp_connected_devices/app/services/token_storage_service.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+
+class TokenStorageError < StandardError; end
+
+class TokenStorageService
+  include SentryLogging
+
+  ##
+  # Uses s3 client to store tokens for the spark token schema
+  # Gracefully handles runtime errors
+  #
+  # @return [boolean]
+  # @raise TokenStorageError
+  def store_tokens(current_user, device_key, tokens_hash)
+    payload_json = unpack_payload(tokens_hash)
+    begin
+      store_locally(current_user, device_key, payload_json) unless Settings.vsp_environment?
+
+      resp = s3_client.put_object(
+        bucket: Settings.dhp.s3.bucket,
+        key: "icn=#{current_user.icn}/device=#{device_key}/tokens.json",
+        body: payload_json
+      )
+
+      resp.etag ? true : false
+    rescue => e
+      raise TokenStorageError, "Error when storing token: #{e.message}"
+    end
+  end
+
+  private
+
+  ##
+  # Package tokens to match Spark's token schema
+  #
+  # @return [json]
+  # @raise TokenStorageError
+  def unpack_payload(tokens_hash)
+    begin
+      tokens_hash.symbolize_keys!
+      unpacked_token = {
+        "payload": {
+          "access_token": tokens_hash[:access_token],
+          "refresh_token": tokens_hash[:refresh_token],
+          "scope": tokens_hash[:scope].gsub(' ', ','),
+          "expires_in": tokens_hash[:expires_in],
+          "received_at": Time.now.utc.strftime('%s%L')
+        }
+      }
+      JSON.dump(unpacked_token)
+    rescue => e
+      raise TokenStorageError, "Error when unpacking token: #{tokens_hash}, #{e}"
+    end
+  end
+
+  def s3_client
+    @s3_client ||= Aws::S3::Client.new(
+      region: Settings.dhp.s3.region,
+      access_key_id: Settings.dhp.s3.aws_access_key_id,
+      secret_access_key: Settings.dhp.s3.aws_secret_access_key
+    )
+  end
+
+  ##
+  # Stores tokens as a read-write file according to the Spark schema
+  # @return [boolean]
+  # @raise [TokenStorageError]
+  def store_locally(current_user, device_key, payload_json)
+    directory_path = "#{::Rails.root}/modules/dhp_connected_devices/local/icn=#{current_user.icn}/device=#{device_key}/tokens.json"
```

Comment:
should the `local` directory be added to `.gitignore`

Reply:
added